File: app/src/main/java/com/hpc/vkd/Downloader.java
package com.hpc.vkd;

import android.annotation.SuppressLint;
import android.os.Environment;
import android.util.Log;

import java.io.File;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

/**
 * Class what makes all the work for downloading the song
 */

public final class Downloader {

    private final static String TAG = Downloader.class.getName();
    static final String FILE_PROVIDER_AUTHORITY = BuildConfig.APPLICATION_ID + ".provider";
    private static AtomicInteger notification_id = new AtomicInteger();
    private static int THREAD_COUNT = 6;
    private static ThreadPoolExecutor NETWORK_POOL_EXECUTOR = new ThreadPoolExecutor(THREAD_COUNT,
            THREAD_COUNT, 0, TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(128));

    private int getNotificationId() {
        return notification_id.incrementAndGet();
    }

    private static Downloader instance;
    private DownloaderListener listener;

    static Downloader getInstance(){
        if (instance == null) {
            instance = new Downloader();
        }
        return instance;
    }

    private Downloader(){
    }

    void subscribe(DownloaderListener listener) {
        this.listener = listener;
    }

    void unsubscribe(){
        this.listener = null;
    }


    @SuppressLint("StaticFieldLeak")
    void getFileNameForUrl(final String fileUrl){
         GetFileNameTask task = new GetFileNameTask() {
            @Override
            protected void onPostExecute(FileName fileName) {
                super.onPostExecute(fileName);
                if (listener != null) {
                    listener.onFileNameReceived(fileUrl, fileName);
                }
            }
        };
        task.execute(fileUrl);
    }

    @SuppressLint("StaticFieldLeak")
    public void downloadFromUrl(final String fileUrl, final FileName fileName) {
        FileParams params = new FileParams(fileUrl,getSaveDir(),fileName);
        Log.i(TAG,this + "getting id");
        final int notificationId = getNotificationId();
        Log.i(TAG,this + "making task for " + fileUrl);
        SaveFileTask task = new SaveFileTask() {
            @Override
            public void doProgressUpdate(FileName name) {
                listener.onLoadStarted(notificationId, name);
            }

            @Override
            public void doPostExecute(File result) {
                listener.onLoadFinished(notificationId, result);
            }
        };
        Log.i(TAG,this + " executing task for " + fileUrl);
        task.executeOnExecutor(NETWORK_POOL_EXECUTOR, params);
    }

    private String getSaveDir() {
        return Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_MUSIC).getAbsolutePath();
    }


}

File: app/src/main/java/com/hpc/vkd/Downloader.kt
package com.hpc.vkd

import android.annotation.SuppressLint
import android.content.ContentValues.TAG
import android.os.Environment
import android.util.Log

import java.io.File
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.ThreadPoolExecutor
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicInteger

/**
 * Class what makes all the work for downloading the song
 */

class Downloader private constructor() {

    private val notificationId: Int
        get() = notification_id.incrementAndGet()
    private var listener: DownloaderListener? = null

    private val saveDir: String
        get() = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_MUSIC).absolutePath

    internal fun subscribe(listener: DownloaderListener) {
        this.listener = listener
    }

    internal fun unsubscribe() {
        this.listener = null
    }


    @SuppressLint("StaticFieldLeak")
    internal fun getFileNameForUrl(fileUrl: String) {
        val task = object : GetFileNameTask() {
            override fun onPostExecute(fileName: FileName) {
                super.onPostExecute(fileName)
                listener?.onFileNameReceived(fileUrl, fileName)
            }
        }
        task.execute(fileUrl)
    }

    @SuppressLint("StaticFieldLeak")
    fun downloadFromUrl(fileUrl: String, fileName: FileName) {
        val params = FileParams(fileUrl, saveDir, fileName)
        Log.i(TAG, this.toString() + "getting id")
        val notificationId = notificationId
        Log.i(TAG, this.toString() + "making task for " + fileUrl)
        val task = object : SaveFileTask() {
            override fun doProgressUpdate(name: FileName) {
                listener?.onLoadStarted(notificationId, name)
            }

            override fun doPostExecute(result: File) {
                listener?.onLoadFinished(notificationId, result)
            }
        }
        Log.i(TAG, this.toString() + " executing task for " + fileUrl)
        task.executeOnExecutor(NETWORK_POOL_EXECUTOR, params)
    }

    companion object {

        private val TAG = Downloader::class.java.name
        internal val FILE_PROVIDER_AUTHORITY = BuildConfig.APPLICATION_ID + ".provider"
        private val notification_id = AtomicInteger()
        private val THREAD_COUNT = 6
        private val NETWORK_POOL_EXECUTOR = ThreadPoolExecutor(THREAD_COUNT,
                THREAD_COUNT, 0, TimeUnit.SECONDS, LinkedBlockingQueue(128))

        private var instance = Downloader()

        internal fun getInstance(): Downloader {
            return instance
        }
    }


}
